File: FinalChallenge/src/final_challenge/scripts/input.cpp
#include <ros/ros.h>
#include <ros/console.h>
#include <std_msgs/Int16.h>
#include <std_msgs/Float32.h>
#include <cmath>
#include "final_challenge/sender.h"

float set_point = 0.0;
double time = 0.0;
float reference = 10;
enum Signal {
    no_signal,
    step_signal,
    square_signal,
    sine_signal
};

void def_signal(int signal) {
  switch (command) {
    case step_signal:
        set_point = time > 0 ? 1 * reference : 0;
        break;
    case square_signal:
        set_point = std::fmod(time*0.5,2) > 1 ? 1 * reference : -1 * reference;
        break;
    case sine_signal:
        set_point = sin(time*2)*reference;
        break;
  }
}

int main(int argc, char *argv[]) {
    ros::init(argc, argv, "sender");
    ros::NodeHandle handler;
    int nodeRate = 100;
    ros::Publisher signalPub = handler.advertise<final_challenge::sender>("/set_point",10);
    ros::Rate rate(nodeRate);
    final::sender set_point_out;
    int signal = 0;
    set_point_out.set_point_data = 0.0;
    while (ros::ok()) {
        ros::param::get("/sp_type", cmd);
        time = ros::Time::now().toSec();
        def_signal(signal);
        set_point_out.set_point_data = set_point;
        signalPub.publish(set_point_out);
        ros::spinOnce();
        rate.sleep();
    }
    return 0;
}

File: FinalChallenge/src/final_challenge/scripts/input.py
import rospy
from std_msgs.msg import Int16, Float32
from math import sin, fmod
from final_challenge.msg import sender

set_point = 0.0
time = 0.0
reference = 10
no_signal = 0
step_signal = 1
square_signal = 2
sine_signal = 3

def def_signal(signal):
    global set_point, time
    if signal == step_signal:
        set_point = 1 * reference if time > 0 else 0
    elif signal == square_signal:
        set_point = 1 * reference if fmod(time * 0.5, 2) > 1 else -1 * reference
    elif signal == sine_signal:
        set_point = sin(time * 2) * reference

def main():
    global set_point, time

    rospy.init_node('sender')
    node_rate = 100
    signal_pub = rospy.Publisher('/set_point', sender, queue_size=10)
    rate = rospy.Rate(node_rate)
    set_point_out = sender()
    signal = 0
    set_point_out.set_point_data = 0.0
    while not rospy.is_shutdown():
        cmd = rospy.get_param("/sp_type")
        time = rospy.get_time()
        def_signal(signal)
        set_point_out.set_point_data = set_point
        signal_pub.publish(set_point_out)
        rate.sleep()

if __name__ == '__main__':
    main()
